Frontend-Backend split

ed255 opened the pull request:
This PR builds on top of https://github.com/privacy-scaling-explorations/halo2/pull/243 which now has become outdated.  I haven't removed the original PR yet because it can be useful for reviewing as it contains a much smaller diff.

The main change in this PR from the previous one is moving all the code into different crates:
- frontend `name = "halo2_frontend"`: the library to write circuits, calculate witness and MockProver
- backend `name = "halo2_backend"`: the library that implements the proof system
- middleware `name = "halo2_middleware"`: the traits and structs used between frontend and backend
- common `name = "halo2_common"`: shared parts between frontend and backend
- halo2_proofs `name = "halo2_proofs"`: lightweight layer + re-exports to make it easy to update a halo2 dependency to this new version, without any API change

Ideally the `common` crate shouldn't exist; but there are many parts of the halo2 implementation that share structs in the frontend and backend.  For example, the Polynomial type is used to calculate commitments in the backend, but also to hold the witness assignment during witness generation in the frontend.  The plan is to move as many things as possible from `common` to either `frontend` or `backend`, ideally reaching a point where `common` disappears.  But that takes some effort so I would request to leave `common` in this PR and iterate it further in other PRs.

Big changes that are not just moving code around in this PR:
- Introduce a new `permutation::keygen::Assembly` called `AssemblyFront` in `common/src/plonk/permutation.rs` to be used by the frontend.  Related to https://github.com/privacy-scaling-explorations/halo2/pull/243#discussion_r1452099723 .  This type now just holds a vector of pairs of cells that need to be copied, without any further processing.  All the mappings and cycles are calculated by the backend.  This also required a change in the MockProver, as it now has direct access to pairs of copy cells, simplifying the error checking.
- Added "Privacy Scaling Explorations team" to the authors of the crates in all `Cargo.toml`.  Should we also add some entry to include the community? 
- Temporarily disabled `thread-safe-region` permutation Assembly in `backend/src/plonk/permutation/keygen.rs`.  I plan to bring this back later.  See https://github.com/privacy-scaling-explorations/halo2/issues/258
- Added `ChallengeMid` in `middleware/src/circuit.rs` and moved original `Challenge` to `common/src/plonk/circuit.rs`.  This is because `Challenge` has a public API which returns `Expression` is is not middleware; so `ChallengeMid` is a lightweight version of `Challenge` that can be converted to `Challenge` to be used in the frontend/backend.
- Added `ColumnMid` in `middleware/src/circuit.rs` which follows the same reasoning as `ChallengeMid`, to not depend on `Expression` in middleware.

Note: I'm not sure what's the best way to review this PR, but I certainly think that looking at the entire diff is not the best approach! The majority of the changes are just moving things around, the actual changes can be reviewed in https://github.com/privacy-scaling-explorations/halo2/pull/243

Missing things
- [x] Remove unnecessary dependencies from each `Cargo.toml`
- [x] Remove the crate `halo2_proofs_rm`.  This is just a copy of halo2 before the crate split, which I'm using for reference while I'm working on this.
- [x] Uncomment all the `allow(unused_code)` and such.  I added these lines to make progress faster; once everything builds and all tests pass, I'll work on that
    - [x] Remove unnecessary imports
- [ ] Bring back `pub(crate)` where necessary.  Since some types are now imported from a different crate, they have become `pub` from `pub(crate)`.  To make my life easier I made a lot of stuff just `pub` while moving things around.  Once types and functions are settled, I need to add `pub(crate)` where necessary. https://github.com/privacy-scaling-explorations/halo2/issues/259
- [x] Configure the lints like in the original `halo2_proofs` for the public facing crates:
```rust
#![cfg_attr(docsrs, feature(doc_cfg))]
// The actual lints we want to disable.
#![allow(clippy::op_ref, clippy::many_single_char_names)]
#![deny(rustdoc::broken_intra_doc_links)]
#![deny(missing_debug_implementations)]
#![deny(missing_docs)]
#![deny(unsafe_code)]
```
- [ ] Add getters (and setters) where necessary.  I made many fields `pub` to make progress faster while I was figuring out where to put things; but the end goal is that user facing types don't expose internals if possible (following the original design of halo2) https://github.com/privacy-scaling-explorations/halo2/issues/259
- [ ] After all reviewing the getters, setters and bringing back `pub(crate)` where appropriate, write the documentation for the new public functions/types that don't have it.  https://github.com/privacy-scaling-explorations/halo2/issues/259
- [x] Fix the tests (in particular, many `use` imports will be broken)
- [x] Fix the benchmarks (I haven't just checked them)
- [x] Fix the code in the docs (the imports will most likely be broken)
- [x] Pass clippy checks
- [ ] Move `poly` to backend https://github.com/privacy-scaling-explorations/halo2/issues/257
- [ ] Bring back `thread-safe-region` permutation Assembly https://github.com/privacy-scaling-explorations/halo2/issues/258
- [ ] Review exposed types/functions in halo2_proofs https://github.com/privacy-scaling-explorations/halo2/issues/260

CPerezz commented:
Sadly seems that Git did not account for ALL the file movements and we have a +30k -5k PR...

It's going to take some time to check this.
Would be important to clearly point to specific commits for each specific change that we do to original Halo2 code.

Such that we can review the commits themselves instead of the entire diff.

ed255 commented:
> Sadly seems that Git did not account for ALL the file movements and we have a +30k -5k PR...

Yeah, the git diff doesn't show all the code moved in the best way, in particular, if a file is split into 2, at most the diff will show a renamed file with deleted lines, and a new file with new lines.  But about the +30k -5k, that was my bad, as in this PR I originally included the old halo2_proofs folder from my previous PR, so there was a lot of duplicated code.  I just cleaned the `halo2_proofs_rm/src` in order to remove duplicated code (there's still benches, examples and tests that need to be put in the right place).  Now the diff is much better at +5.7k - 3k :D

> It's going to take some time to check this. Would be important to clearly point to specific commits for each specific change that we do to original Halo2 code.
> 
> Such that we can review the commits themselves instead of the entire diff.

That's a good point; so far I added a list of halo2 code changes on top of my first PR that are included here.  I'll also write down the commits where that happens!

han0110 commented on halo2_proofs/src/plonk/prover.rs:
```diff
@@ -56,674 +33,38 @@ where
+    let mut challenges = HashMap::new();
+    let phases = prover.phases.clone();
+    for phase in &phases {
+        println!("DBG phase {}", phase.0);
```
leftover

ed255 replied:
Resolved in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/134143c618f41aea4805fd0cff14962c819397f1

han0110 commented on halo2_proofs/tests/frontend_backend_split.rs:
```diff
@@ -0,0 +1,612 @@
+        mut layouter: impl Layouter<F>,
+    ) -> Result<(), Error> {
+        // 2 queries from first gate, 3 for permutation argument, 1 for multipoen, 1 for off-by-one
+        //   errors, 1 for off-by-two errors?
```
> 1 for off-by-two errors

It should be 1 extra row for the last row of grand product poly to check if it products to `1`.

ed255 replied:
Thanks for clarifying this point!

ed255 replied:
Resolved in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/134143c618f41aea4805fd0cff14962c819397f1

han0110 reviewed (approved):
Overall LGTM! Nice work! Just some leftovers needed to be cleaned up.

ed255 commented on backend/src/plonk/prover.rs:
```diff
@@ -0,0 +1,743 @@
+            .chain(vanishing.open(x));
+
+        let prover = P::new(params);
+        println!("DBG create_proof");
```
Resolved in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/134143c618f41aea4805fd0cff14962c819397f1

ed255 commented on common/src/plonk/circuit.rs:
```diff
@@ -1536,66 +1465,432 @@ impl<F: Field> Gate<F> {
+    }
+}
+
+/*
```
Resolved in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/134143c618f41aea4805fd0cff14962c819397f1

ed255 commented on backend/src/plonk/keygen.rs:
```diff
@@ -0,0 +1,189 @@
+}
+
+/// Generate a `VerifyingKey` from an instance of `CompiledCircuit`.
+pub fn keygen_vk_v2<'params, C, P>(
```
That's a good point!  I forgot to bring back the discussion on naming.  I'll create an issue to have this discussion separately and address it in a future PR: https://github.com/privacy-scaling-explorations/halo2/issues/263

ed255 commented on middleware/src/circuit.rs:
```diff
@@ -0,0 +1,354 @@
+    // /// Return Expression
+    // pub fn expr<F: Field>(&self) -> ExpressionMid<F> {
+    //     ExpressionMid::Challenge(*self)
+    // }
```
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/7eda496db87a8b41d079d6e5b87c42cde4434c93

ed255 commented on Cargo.toml:
```diff
@@ -2,4 +2,8 @@
+    "frontend",
+    "middleware",
+    "backend",
+    "common",
```
Thanks for this!  I wasn't aware this was common :) 
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/7eda496db87a8b41d079d6e5b87c42cde4434c93

ed255 commented:
> The tmp removal of the `thread-safe-region` makes sense. We should anyways fill an issue to re-address it after this is merged.

I created issues for all the tasks that need to be done after this is merged, I hope I didn't forget anything!  This is the issue for the `thread-safe-region`: https://github.com/privacy-scaling-explorations/halo2/issues/258

ed255 commented on backend/src/plonk.rs:
```diff
@@ -0,0 +1,466 @@
+    selectors: Vec<Vec<bool>>,
+    // TODO: Use setter/getter
+    /// Whether selector compression is turned on or not.
+    pub compress_selectors: bool,
```
Updated with a link to the issue https://github.com/privacy-scaling-explorations/halo2/issues/259

ed255 replied:
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on backend/src/plonk.rs:
```diff
@@ -0,0 +1,466 @@
+    pub fn from_bytes<ConcreteCircuit: Circuit<C::Scalar>>(
+        mut bytes: &[u8],
+        format: SerdeFormat,
+        #[cfg(feature = "circuit-params")] params: ConcreteCircuit::Params,
```
I know!  If I write this:
```rust
        #[cfg(feature = "circuit-params")]
        params: ConcreteCircuit::Params,
```
cargo fmt turns it back to:
```rust
       #[cfg(feature = "circuit-params")] params: ConcreteCircuit::Params,
```

ed255 commented on backend/src/plonk/keygen.rs:
```diff
@@ -0,0 +1,189 @@
+    .build_pk(params, &vk.domain, &cs.permutation.clone().into());
+
+    // Compute l_0(X)
+    // TODO: this can be done more efficiently
```
I don't know either :laughing: !  Let me find the source of this...
https://github.com/privacy-scaling-explorations/halo2/commit/c97da352ee1350ae42b0f8789eb177832f3b6cd2#diff-9ee64a77d20a4f698af319edd60ee8d1b39db759d1abf9c5a4bd11cbe83b41edR239

ed255 replied:
https://github.com/privacy-scaling-explorations/halo2/blob/73408a140737d8336490452193b21f5a7a94e7de/halo2_proofs/src/plonk/keygen.rs#L360-L365

ed255 commented on backend/src/plonk/permutation/prover.rs:
```diff
@@ -0,0 +1,328 @@
+
+pub(crate) struct Evaluated<C: CurveAffine> {
+    constructed: Constructed<C>,
+}
```
I agree!  Nevertheless I'm not familiar with this code, it's just ported from https://github.com/privacy-scaling-explorations/halo2/blob/73408a140737d8336490452193b21f5a7a94e7de/halo2_proofs/src/plonk/permutation/prover.rs#L21

I'll create an issue for that

ed255 replied:
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on backend/src/plonk/prover.rs:
```diff
@@ -0,0 +1,743 @@
+        params: &'params Scheme::ParamsProver,
+        pk: &'a ProvingKey<Scheme::Curve>,
+        // TODO: If this was a vector the usage would be simpler
+        instances: &[&[&[Scheme::Scalar]]],
```
yes! https://github.com/privacy-scaling-explorations/halo2/issues/265

ed255 replied:
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/bbf034a1f1207319ab0318e3ecfabe92392c2116

ed255 commented on backend/src/plonk/prover.rs:
```diff
@@ -0,0 +1,743 @@
+        transcript: &'a mut T,
+    ) -> Result<Self, Error>
+    // TODO: Can I move this `where` to the struct definition?
+    where
```
I will remove the comment as I didn't find a way to do it.

ed255 replied:
I think we can just remove the comment without issue, because it's a tiny thing.
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on backend/src/plonk/shuffle/prover.rs:
```diff
@@ -0,0 +1,250 @@
+
+pub(in crate::plonk) struct Evaluated<C: CurveAffine> {
+    constructed: Committed<C>,
+}
```
I didn't find the duplicate, could you show where that is?  Maybe you're thinking about the `premutation/prover.rs` types?  They have some similarities but are actually different.

ed255 replied:
I think trying to reuse structs in this particular part may make things more complicated.  On the other hand, the shuffle, copy constraint and lookup arguments use a grand product and challenges, so I think halo2 could have an internal API for these kinds of arguments so deduplicate code.  I think @kilic told me this idea, I'm not sure if there's an issue for that.

ed255 commented on common/src/circuit/value.rs:
```diff
@@ -64,7 +65,7 @@ impl<V> Value<V> {
 
     /// ONLY FOR INTERNAL CRATE USAGE; DO NOT EXPOSE!
-    pub(crate) fn into_option(self) -> Option<V> {
+    pub fn into_option(self) -> Option<V> {
```
This function was only used twice, and it's dangerous, so I've just removed it and did a workaround.

ed255 replied:
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on common/src/lib.rs:
```diff
@@ -0,0 +1,24 @@
+#![allow(clippy::op_ref, clippy::many_single_char_names)]
+#![deny(rustdoc::broken_intra_doc_links)]
+// #![deny(missing_debug_implementations)] // TODO: Uncomment
+// #![deny(missing_docs)] // TODO: Uncomment
```
I'll remove those lines with the following reasoning:
Common types/functions that are currently shared by the frontend and backend so they are `pub` but are meant for internal usage only; so they should follow the same requirements as `pub(crate)` types/functions.  Hopefully in future work we can move this types to the frontend/backend and they can truly become `pub(crate)`.

ed255 replied:
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on common/src/lib.rs:
```diff
@@ -0,0 +1,24 @@
+pub use halo2curves;
+pub mod multicore;
+pub mod plonk;
+// TODO: Try to move this to backend and use a lightweight Polynomial type in the frontend
```
Yes! I already opened the issue, I'll cite it in the comment https://github.com/privacy-scaling-explorations/halo2/issues/257

ed255 replied:
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on backend/src/lib.rs:
```diff
@@ -0,0 +1,12 @@
+#![allow(unused_imports)] // TODO: Remove
```
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on backend/src/plonk.rs:
```diff
@@ -0,0 +1,466 @@
+// TODO: Import propertly instead of reimporting from inside
+pub use halo2_common::plonk::{
+    ChallengeBeta, ChallengeGamma, ChallengeTheta, ChallengeX, ChallengeY, Error,
+};
```
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on backend/src/plonk.rs:
```diff
@@ -0,0 +1,466 @@
+#[derive(Clone, Debug)]
+pub struct ProvingKey<C: CurveAffine> {
+    // TODO: Add getter
+    pub vk: VerifyingKey<C>,
```
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on backend/src/plonk/prover.rs:
```diff
@@ -0,0 +1,743 @@
+
+        let domain = &pk.vk.domain;
+
+        // TODO: Name this better
```
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on backend/src/plonk/prover.rs:
```diff
@@ -0,0 +1,743 @@
+        };
+        if phase != current_phase.0 {
+            panic!("TODO: Return Error instead. Committing invalid phase");
+        }
```
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on backend/src/plonk/prover.rs:
```diff
@@ -0,0 +1,743 @@
+                    )
+                })
+                .collect();
+            // dbg!(&advice_evals);
```
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on common/src/lib.rs:
```diff
@@ -0,0 +1,24 @@
+pub use helpers::SerdeFormat;
+
+// TODO: Everything that is moved from this crate to frontend or backend should recover the
+// pub status whenever possible.
```
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on common/src/plonk/error.rs:
```diff
@@ -3,7 +3,10 @@ use std::fmt;
+use crate::plonk::circuit::Column;
+use halo2_middleware::circuit::Any;
+
+// TODO: Consider splitting this Error into a frontend and backend version?
```
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on middleware/Cargo.toml:
```diff
@@ -0,0 +1,42 @@
+"""
+license = "MIT OR Apache-2.0"
+repository = "TODO"
+documentation = "TODO"
```
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163

ed255 commented on backend/src/plonk.rs:
```diff
@@ -0,0 +1,466 @@
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "unexpected version byte",
+            ));
```
Opened issue here https://github.com/privacy-scaling-explorations/halo2/issues/267

ed255 commented on backend/src/plonk.rs:
```diff
@@ -0,0 +1,466 @@
+            cs,
+            cs_degree,
+            // Temporary, this is not pinned.
+            transcript_repr: C::Scalar::ZERO,
```
This should be OK.  This is a workaround from a cyclic dependency:
- The Vk includes a hash of parts of the Vk
- The method that serializes the parts of the Vk that are used in the hash, is implemented on top of the VerifyingKey type

So the solution is create a Vk with that value at 0, format the pinned version and use it as a serialization of the parts of the Vk for hashing, and then add that hash result to the Vk.

ed255 commented on backend/src/plonk.rs:
```diff
@@ -0,0 +1,466 @@
+    {
+        let scalar_len = C::Scalar::default().to_repr().as_ref().len();
+        self.vk.bytes_length(format)
+            + 12
```
Totally agree.  Opened issue https://github.com/privacy-scaling-explorations/halo2/issues/268

ed255 commented on backend/src/plonk/keygen.rs:
```diff
@@ -0,0 +1,189 @@
+
+pub(crate) fn create_domain<C, ConcreteCircuit>(
+    k: u32,
+    #[cfg(feature = "circuit-params")] params: ConcreteCircuit::Params,
```
Resolved following https://github.com/privacy-scaling-explorations/halo2/pull/254#discussion_r1472442487

ed255 commented on middleware/src/lib.rs:
```diff
@@ -0,0 +1,108 @@
+
+pub use ff;
+
+// TODO: Remove with permutation::Argument simplification
```
yeah! This was infact duplicated, this code was already in `common` and doesn't belong to middleware.

ed255 replied:
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/9927c824927cf82acf7a7c45848a7f86f85a7416

ed255 commented on middleware/src/permutation.rs:
```diff
@@ -0,0 +1,20 @@
+use crate::circuit::ColumnMid;
+
+// TODO: Dedup with other Cell definition, or move this to a higher level
```
Can't be deduped easily because the other `Cell` has region info, and this one is in the middleware where the concept or Region doesn't apply.  I'll move it to a higher level instead.

ed255 replied:
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/9927c824927cf82acf7a7c45848a7f86f85a7416

ed255 commented on middleware/src/lib.rs:
```diff
@@ -0,0 +1,108 @@
+    }
+}
+
+// TODO: Remove with permutation::Argument simplification
```
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/9927c824927cf82acf7a7c45848a7f86f85a7416

ed255 commented on middleware/src/circuit.rs:
```diff
@@ -0,0 +1,354 @@
+    pub cs: ConstraintSystemV2Backend<F>,
+}
+
+// TODO: The query_cell method is only used in the frontend, which uses Expression.  By having this
```
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/9927c824927cf82acf7a7c45848a7f86f85a7416

ed255 commented on common/src/plonk/permutation.rs:
```diff
@@ -0,0 +1,119 @@
+    pub n: usize,
+    pub columns: Vec<Column<Any>>,
+    pub copies: Vec<(Cell, Cell)>,
+}
```
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/9927c824927cf82acf7a7c45848a7f86f85a7416

ed255 commented:
> @ed255 Why we need middleware plonk module if is not used by CompiledCircuit? ( AFAIU the main goal of middleware is to hold a `CompiledCircuit` )

The `plonk::Assigned` type is used to encode the witness that the frontend passes to the backend via the backend method: 
https://github.com/privacy-scaling-explorations/halo2/blob/25666a4c52dfd97f7a6a89e43b0ed903743ba88d/halo2_backend/src/plonk/prover.rs#L258-L267

The reason for having `Vec<Assigned<Scheme::Scalar>>` instead of `Vec<Scheme::Scalar>` is that currently the halo2 frontend sometimes generates divisions or inverted field elements as witness; and it's more efficient to do the inversion in a batch; and that's currently being done in the backend.  It could also be done in the frontend, and then we would be able to remove the `Assigned` from the middleware.  Perhaps that would be better!
